### PR TITLE
Sanity Tests

### PR DESCRIPTION
**Pull request by maennchen**
## Setup:

Create File `test/secret.yml`

```yaml
CreateVolumeSecret:
  projectId: project-id
  key: |
    key json
DeleteVolumeSecret:
  projectId: project-id
  key: |
    key json
ControllerPublishVolumeSecret:
  projectId: project-id
  key: |
    key json
ControllerUnpublishVolumeSecret:
  projectId: project-id
  key: |
    key json
NodeStageVolumeSecret:
  projectId: project-id
  key: |
    key json
NodePublishVolumeSecret:
  projectId: project-id
  key: |
    key json
ControllerValidateVolumeCapabilitiesSecret:
  projectId: project-id
  key: |
    key json
```

## Run 

```bash
invoke sanity-test
```

## TODO

1. [x] CI Integration
1. [x] Set `TestSecrets` GitHub Workflow Secret (@ofek)
1. [x] A Docker Setup (easier local setup)
1. [x] An `invoke` command that builds the docker image and runs the test
1. [x] Fix [`NodeUnpublishVolume test for when the volume is missing` Test](https://github.com/kubernetes-csi/csi-test/commit/5b292b5663789dc1f134eb119756d41c7b46336a)

## `NodeUnpublishVolume test for when the volume is missing` Test

This test mandates, that there should be a `NotFound` status if the volume does not exist. There is however no `Secret` supplied to the function so that it could actually be checked.

Since the function is supposed to be idempotent (should not fail when called a second time), the presence of the `target` can also not be used to detect the condition.

```
• Failure [0.005 seconds]
Node Service
/go/pkg/mod/github.com/kubernetes-csi/csi-test/v3@v3.1.0/pkg/sanity/tests.go:44
  NodeUnpublishVolume
  /go/pkg/mod/github.com/kubernetes-csi/csi-test/v3@v3.1.0/pkg/sanity/node.go:396
    should fail when the volume is missing [It]
    /go/pkg/mod/github.com/kubernetes-csi/csi-test/v3@v3.1.0/pkg/sanity/node.go:423

    Expected an error to have occurred.  Got:
        <nil>: nil

    /go/pkg/mod/github.com/kubernetes-csi/csi-test/v3@v3.1.0/pkg/sanity/node.go:431
```

### CODE REVIEW

**Comment by maennchen** [comment]
@ofek I'm making some progress. I'll push an update soon.

**Comment by maennchen** [comment]
This needs three more things to be nice:

1. CI Integration
1. A Docker Setup (easier local setup)
1. An `invoke` comand that builds the docker image and runs the test

**Comment by maennchen** [comment]
@ofek This is my best idea for the `NotFound` problem:

```diff
diff --git a/pkg/driver/driver.go b/pkg/driver/driver.go
index d45b5ce..ed8d1ce 100644
--- a/pkg/driver/driver.go
+++ b/pkg/driver/driver.go
@@ -13,21 +13,23 @@ import (
 )
 
 type GCSDriver struct {
-	name       string
-	nodeName   string
-	endpoint   string
-	mountPoint string
-	version    string
-	server     *grpc.Server
+	name                 string
+	nodeName             string
+	endpoint             string
+	mountPoint           string
+	version              string
+	server               *grpc.Server
+	encounteredVolumeIds map[string]bool
 }
 
 func NewGCSDriver(name, node, endpoint string) (*GCSDriver, error) {
 	return &GCSDriver{
-		name:       name,
-		nodeName:   node,
-		endpoint:   endpoint,
-		mountPoint: BucketMountPath,
-		version:    driverVersion,
+		name:                 name,
+		nodeName:             node,
+		endpoint:             endpoint,
+		mountPoint:           BucketMountPath,
+		version:              driverVersion,
+		encounteredVolumeIds: map[string]bool{},
 	}, nil
 }
 
diff --git a/pkg/driver/node.go b/pkg/driver/node.go
index b03db95..c481635 100644
--- a/pkg/driver/node.go
+++ b/pkg/driver/node.go
@@ -64,6 +64,10 @@ func (driver *GCSDriver) NodePublishVolume(ctx context.Context, req *csi.NodePub
 		return nil, status.Errorf(codes.NotFound, "Bucket %s does not exist", bucketName)
 	}
 
+	// This is an on purpose introduced memory leak to solve
+	// https://github.com/kubernetes-csi/csi-test/commit/5b292b5663789dc1f134eb119756d41c7b46336a#commitcomment-38586185
+	driver.encounteredVolumeIds[volumeId] = true
+
 	var extraFlags = []string{}
 	if flags, exists := options["flags"]; exists {
 		parsedFlags := strings.Fields(flags)
@@ -105,6 +109,11 @@ func (driver *GCSDriver) NodeUnpublishVolume(ctx context.Context, req *csi.NodeU
 		return nil, status.Error(codes.InvalidArgument, "Target path missing in request")
 	}
 
+	_, ok := driver.encounteredVolumeIds[volumeId]
+	if !ok {
+		return nil, status.Error(codes.NotFound, "Volume unknown")
+	}
+
 	notMount, err := CheckMount(targetPath)
 	if err != nil || notMount {
 		return &csi.NodeUnpublishVolumeResponse{}, nil
```

**Comment by maennchen** [comment]
@ofek 

1. How do you want to continue with the failing test for now? I would recommend merging even with one failed test and creating a separate issue. When the discussion upstream is resolved, we'll react accordingly.
1. To get all test but one passing, we'll have to add the `TestSecrets` Secret.

**Comment by ofek** [comment]
Is there any way to skip it e.g. so the badge will be green?

**Comment by maennchen** [comment]
@ofek I added `continue-on-error`.

**Comment by maennchen** [comment]
@ofek (Still add the secrets and test that only one test fails :) )

**Comment by ofek** [comment]
Any recommendation on how to construct the secrets file?

**Comment by maennchen** [comment]
@ofek I had no idea other than manual copy / paste. If it took separate files / a json or something, I would‘ve generated the file automatically. But like this I‘ve no better & reasonable idea.

**Comment by ofek** [comment]
Capture

**Comment by maennchen** [comment]
@ofek What are the exact errors? (Should be further up in the output)

**Comment by ofek** [comment]
Capture0

Capture0

Capture0

**Comment by maennchen** [comment]
@ofek: Classic „Works on my Machine“.

Can you add some log statements to debug why the bucket does not exist?

Commenting out the DeleteVolume GCS Call should also leave the buckets so you can see if they were created.

**Comment by maennchen** [comment]
@ofek Did you use the same service account key for all operations or did you use two separate ones?

If two separate: Does it work when the check if the volume exists is commented out?

**Comment by ofek** [comment]
Used 2

**Comment by maennchen** [comment]
@ofek Does it work with storage admin for all use-cases?

It might be the attrs call to detect if the bucket exists.

**Comment by ofek** [comment]
Yup switching NodePublishVolumeSecret & NodeStageVolumeSecret to the other key worked

**Comment by maennchen** [comment]
@ofek There‘s two ways on how to solve this then I guess:

1. Chaning the docs to mention that the service account needs more permissions.
1. Finding a better way to detect if a bucket exists. (I think though that the more of the CSI spec is supported, the bigger the need to read attrs / labels.)

**Comment by ofek** [comment]
I'm fine with 1. if

1. it only affects dynamic provisioning
1. we give a gcloud command
1. maybe opt-out flag to disable whatever is failing?

**Comment by maennchen** [comment]
@ofek I‘ll try to come up with a better check if the bucket exists that doesn’t need more permissions...

**Comment by maennchen** [comment]
@ofek There's now two ways on how to check if a bucket exists:

* `Controller` - using `Attrs` since controller needs to be able to read / write Attrs.
* `Node` - using `Objects` since node needs to be able to see Objects.

**Comment by ofek** [comment]
It works!

Capture0

File: pkg/util/common.go [hunk 1]
```diff
@@ -111,8 +112,8 @@ func BucketName(volumeId string) string {
 	// return volumeId
 	var crc32Hash = crc32.ChecksumIEEE([]byte(volumeId))
 
-	if len(volumeId) > 32 {
-		volumeId = volumeId[0:32]
+	if len(volumeId) > 48 {
```
Comment (ofek) [inline pkg/util/common.go:115]: What's this number?

File: tasks/test.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        f'docker build . --tag {image} '
+        f'--build-arg version={version} '
+        f'--build-arg global_ldflags="{global_ldflags}" '
+        f'--build-arg gcsfuse_version="{gcsfuse}" '
+        f'--target test',
+        echo=True,
+    )
+
+    ctx.run(
+        f'docker run '
```
Comment (ofek) [inline tasks/test.py:30]:
```suggestion
        f'docker run --rm '
```

File: pkg/util/common.go (hunk 1 above)
Comment (maennchen) [reply pkg/util/common.go:115]: @ofek 16 characters for the hash leaves 48 for the name. (limit: 64)

Comment (maennchen) [reply pkg/util/common.go:115]: (Saw that I had a little low number before here)

**Pull request merged**